tmail-backend/tmail-third-party/openpaas/src/test/java/com/linagora/tmail/OpenPaasAmqpForwardAttributeIntegrationTest.java

```diff
@@ -0,0 +1,322 @@
+import org.apache.james.modules.protocols.ImapGuiceProbe;
+import org.apache.james.modules.protocols.SmtpGuiceProbe;
+import org.apache.james.transport.mailets.ContactExtractor;
+import org.apache.james.transport.matchers.All;
+import org.apache.james.transport.matchers.SMTPAuthSuccessful;
+import org.apache.james.utils.DataProbeImpl;
+import org.apache.james.utils.SMTPMessageSender;
+import org.apache.james.utils.TestIMAPClient;
+import org.apache.mailet.base.test.FakeMail;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.api.io.TempDir;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.Provides;
+import com.google.inject.util.Modules;
+import com.linagora.tmail.configuration.OpenPaasConfiguration;
+import com.linagora.tmail.james.jmap.contact.InMemoryEmailAddressContactSearchEngineModule;
+import com.linagora.tmail.mailet.OpenPaasAmqpForwardAttribute;
+import com.linagora.tmail.mailet.OpenPaasAmqpForwardAttributeConfig;
+import com.rabbitmq.client.BuiltinExchangeType;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.GetResponse;
+
+class OpenPaasAmqpForwardAttributeIntegrationTest {
+    private static final boolean LOAD_OPENPAAS_MODULE = true;
+
+    public static final String SENDER = "sender@" + DEFAULT_DOMAIN;
+    public static final String TO = "to@" + DEFAULT_DOMAIN;
+    public static final String EXTRACT_ATTRIBUTE = "ExtractedContacts";
+
+    @RegisterExtension
+    public static RabbitMQExtension rabbitMQExtension = RabbitMQExtension.singletonRabbitMQ()
+        .isolationPolicy(RabbitMQExtension.IsolationPolicy.WEAK);
+
+    @RegisterExtension
+    public TestIMAPClient testIMAPClient = new TestIMAPClient();
+    @RegisterExtension
+    public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
+
+    private TemporaryJamesServer jamesServer;
+    private Connection rabbitMQConnection;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        ConnectionFactory connectionFactory = new ConnectionFactory();
+        connectionFactory.setUri(rabbitMQExtension.getRabbitMQ().amqpUri());
+
+        rabbitMQConnection = connectionFactory.newConnection();
+    }
+
+    @AfterEach
+    void tearDown() {
+        if (jamesServer != null) {
+            jamesServer.shutdown();
+        }
+        if (rabbitMQConnection != null) {
+            try {
+                rabbitMQConnection.close();
+            } catch (Exception ignored) {
+                // Ignored
+            }
+        }
+    }
+
+    private void buildJameServer(File temporaryFolder, MailetConfiguration amqpForwardAttributeMailet, boolean loadOpenPaasModule) throws Exception {
```

Comment:
Or.... we could have several test classes each one with the right james server set up

Or... We could rely on nested classes

But such a long setup method with so much possible states is evil.

Reply:
Isn't this too much for three test cases? It would make sense only if there were no alternatives, but I believe we can hide the evil states through refactoring. I'm thinking of creating two methods: buildJamesServerWithOpenPaas and buildJamesServerWithoutOpenPaas.

Reply:
Or... We could rely on nested classes